src/transformers/models/jamba/modeling_jamba.py

```diff
@@ -0,0 +1,2127 @@
+            else:
+                # we don't have cache for this layer, initialize it with zeros
+                batch_size = hidden_states.shape[0]
+                cache_params.conv_states[self.layer_idx] = torch.zeros(
+                    batch_size,
+                    self.intermediate_size,
+                    self.conv_kernel_size,
+                    device=hidden_states.device,
+                    dtype=hidden_states.dtype,
+                )
+                cache_params.ssm_states[self.layer_idx] = torch.zeros(
+                    batch_size,
+                    self.intermediate_size,
+                    self.ssm_state_size,
+                    device=hidden_states.device,
+                    dtype=hidden_states.dtype,
+                )
+        else:
+            cache_params = None
+
+        res = self.mixer_forward(hidden_states, cache_params)
+
+        if past_key_value is not None:
+            past_key_value.update(
+                # add dummy seqlen dim (dim=2) to match the number of dimensions of the attention cache
+                cache_params.conv_states[self.layer_idx].unsqueeze(2),
+                cache_params.ssm_states[self.layer_idx].unsqueeze(2),
+                self.layer_idx,
+            )
+
+        return res, past_key_value
+
+
+class JambaMLP(nn.Module):
+    def __init__(self, config: JambaConfig):
+        super().__init__()
+        self.ffn_dim = config.intermediate_size
+        self.hidden_dim = config.hidden_size
+
+        self.gate_proj = nn.Linear(self.hidden_dim, self.ffn_dim, bias=False)
+        self.down_proj = nn.Linear(self.ffn_dim, self.hidden_dim, bias=False)
+        self.up_proj = nn.Linear(self.hidden_dim, self.ffn_dim, bias=False)
+
+        self.act_fn = ACT2FN[config.hidden_act]
+
+    def forward(self, x):
+        return self.down_proj(self.act_fn(self.gate_proj(x)) * self.up_proj(x))
+
+
+# Adapted from transformers.models.mixtral.modeling_mixtral.MixtralSparseMoeBlock with Mistral->Jamba
+class JambaSparseMoeBlock(nn.Module):
+    """
+    This implementation is
+    strictly equivalent to standard MoE with full capacity (no
+    dropped tokens). It's faster since it formulates MoE operations
+    in terms of block-sparse operations to accomodate imbalanced
+    assignments of tokens to experts, whereas standard MoE either
+    (1) drop tokens at the cost of reduced performance or (2) set
+    capacity factor to number of experts and thus waste computation
+    and memory on padding.
+    """
+
+    def __init__(self, config: JambaConfig, num_experts: int, num_experts_per_tok: int):
+        super().__init__()
+        self.hidden_dim = config.hidden_size
+        self.ffn_dim = config.intermediate_size
+
+        #   these values are decided on runtime depending on the layer index
+        self.num_experts = num_experts
+        self.top_k = num_experts_per_tok
+
+        if num_experts > 1:
```

Comment:
if num expert is one we should not use a MoE block IMO, should be checked outside this function!

Reply:
In `switch_transformers` we used a `SwitchTransformersSparseMLP` vs a `JambaMLP`. 
Not super fan of having two code path, would be nice to just have two class.

Reply:
`JambaSparseMoeBlock`'s init can thus be copied from `MixtralSparseMoeBlock` with renaming (self.router = gate`